File: 2024/5/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8")
  .split("\n\n");

const rules = input[0].split("\n").map((a) => a.split("|").map(Number));
const updates = input[1].split("\n").map((a) => a.split(",").map(Number));

class Rule {
  constructor(num) {
    this.num = num;
    this.afterNum = new Set();
  }

  addRule(n) {
    this.afterNum.add(n);
  }
}

const ruleMap = new Map();

for (const [n1, n2] of rules) {
  if (!ruleMap.has(n1)) {
    ruleMap.set(n1, new Rule(n1));
  }
  ruleMap.get(n1).addRule(n2);
}

function sortingFunction(a, b) {
  if (ruleMap.has(a) && ruleMap.get(a).afterNum.has(b)) return -1;
  if (ruleMap.has(b) && ruleMap.get(b).afterNum.has(a)) return 1;
  return 0;
}

let part1 = 0;
let part2 = 0;

for (const update of updates) {
  const order = new Map();
  for (let i = 0; i < update.length; i++) {
    order.set(update[i], i);
  }
  let isCorrectOrder = true;
  for (let i = 0; i < update.length; i++) {
    if (ruleMap.has(update[i])) {
      const rule = ruleMap.get(update[i]);
      for (const n of rule.afterNum) {
        if (order.has(n) && order.get(n) < i) {
          isCorrectOrder = false;
          break;
        }
      }
    }
    if (!isCorrectOrder) break;
  }
  if (isCorrectOrder) {
    part1 += update[Math.floor(update.length / 2)];
  } else {
    update.sort(sortingFunction);
    part2 += update[Math.floor(update.length / 2)];
  }
}

console.log(part1);
console.log(part2);

File: 2024/5/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"slices"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	puzzleInput := strings.Split((string(data)), "\n\n")
	rules, pageUpdates := parseInput(puzzleInput)
	part1Answer, incorrectPages := part1(rules, pageUpdates)
	fmt.Println(part1Answer)
	fmt.Println(part2(rules, incorrectPages))
}

func convertToInts(data []string) [][]int {
	res := make([][]int, len(data))
	numRegex := regexp.MustCompile(`\d+`)
	for i, s := range data {
		nums := numRegex.FindAllString(s, -1)
		for _, num := range nums {
			if n, err := strconv.Atoi(num); err != nil {
				panic(err)
			} else {
				res[i] = append(res[i], n)
			}
		}
	}
	return res
}

func parseInput(data []string) (map[int]map[int]struct{}, [][]int) {
	ruleInput := strings.Split(data[0], "\n")
	pageInput := strings.Split(data[1], "\n")
	ruleList := convertToInts(ruleInput)
	rules := make(map[int]map[int]struct{})
	for _, row := range ruleList {
		if _, ok := rules[row[0]]; !ok {
			rules[row[0]] = make(map[int]struct{})
		}
		rules[row[0]][row[1]] = struct{}{}
	}
	return rules, convertToInts(pageInput)
}

func isCorrectOrder(pages []int, rules map[int]map[int]struct{}) bool {
	pageMap := make(map[int]int, len(pages))
	for i, j := range pages {
		pageMap[j] = i
	}
	for i, page1 := range pages {
		if rule, ok := rules[page1]; ok {
			for page2 := range rule {
				if j, ok := pageMap[page2]; ok && j < i {
					return false
				}
			}
		}
	}
	return true
}

func part1(rules map[int]map[int]struct{}, pageUpdates [][]int) (int, [][]int) {
	res := 0
	incorrectPages := make([][]int, 0, len(pageUpdates))
	for _, pages := range pageUpdates {
		if isCorrectOrder(pages, rules) {
			res += pages[len(pages)/2]
		} else {
			incorrectPages = append(incorrectPages, pages)
		}
	}
	return res, incorrectPages
}

func part2(rules map[int]map[int]struct{}, pageUpdates [][]int) int {
	res := 0
	for _, pages := range pageUpdates {
		slices.SortFunc(pages, func(a, b int) int {
			if _, ok := rules[a]; ok {
				if _, ok := rules[a][b]; ok {
					return -1
				}
			}
			if _, ok := rules[b]; ok {
				if _, ok := rules[b][a]; ok {
					return 1
				}
			}
			return 0
		})
		res += pages[len(pages)/2]
	}
	return res
}
